src/Common/tests/System/Diagnostics/DebuggerAttributes.cs

```diff
@@ -0,0 +1,140 @@
+                .Where(a => a.AttributeType == typeof(DebuggerTypeProxyAttribute))
+                .ToArray();
+            if (attrs.Length != 1)
+            {
+                throw new InvalidOperationException(
+                    string.Format("Expected one DebuggerTypeProxyAttribute on {0}.", obj));
+            }
+            var cad = (CustomAttributeData)attrs[0];
+
+            // Get the proxy type.  As written, this only works if the proxy and the target type
+            // have the same generic parameters, e.g. Dictionary<TKey,TValue> and Proxy<TKey,TValue>.
+            // It will not work with, for example, Dictionary<TKey,TValue>.Keys and Proxy<TKey>,
+            // as the former has two generic parameters and the latter only one.
+            Type proxyType = cad.ConstructorArguments[0].ArgumentType == typeof(Type) ?
+                (Type)cad.ConstructorArguments[0].Value :
+                Type.GetType((string)cad.ConstructorArguments[0].Value);
+            var genericArguments = obj.GetType().GenericTypeArguments;
+            if (genericArguments.Length > 0)
+            {
+                proxyType = proxyType.MakeGenericType(genericArguments);
+            }
+
+            // Create an instance of the proxy type, and make sure we can access all of the instance properties 
+            // on the type without exception
+            object proxyInstance = Activator.CreateInstance(proxyType, obj);
+            foreach (var pi in proxyInstance.GetType().GetTypeInfo().DeclaredProperties)
+            {
+                pi.GetValue(proxyInstance, null);
+            }
+        }
+
+        internal static void ValidateDebuggerDisplayReferences(object obj)
+        {
+            // Get the DebuggerDisplayAttribute for obj
+            var attrs = 
+                obj.GetType().GetTypeInfo().CustomAttributes
+                .Where(a => a.AttributeType == typeof(DebuggerDisplayAttribute))
+                .ToArray();
+            if (attrs.Length != 1)
+            {
+                throw new InvalidOperationException(
+                    string.Format("Expected one DebuggerDisplayAttribute on {0}.", obj));
+            }
+            var cad = (CustomAttributeData)attrs[0];
+
+            // Get the text of the DebuggerDisplayAttribute
+            string attrText = (string)cad.ConstructorArguments[0].Value;
+
+            // Parse the text for all expressions
+            var references = new List<string>();
+            int pos = 0;
+            while (true)
+            {
+                int openBrace = attrText.IndexOf('{', pos);
+                if (openBrace < pos) break;
+                int closeBrace = attrText.IndexOf('}', openBrace);
+                if (closeBrace < openBrace) break;
+
+                string reference = attrText.Substring(openBrace + 1, closeBrace - openBrace - 1).Replace(",nq", "");
+                pos = closeBrace + 1;
+
+                references.Add(reference);
```

Comment:
I did this on purpose.  It's considered bad practice to use arbitrary expressions (in particular because the expression is evaluated by the language provider currently in use while debugging, and syntax of course varies by language), and in fact that's another bug we've fixed in our debugger attributes a few times (e.g. #1025).  If we'd had this test in place, it would have caught it.  The best practice is to use simple fields or properties, and if you need a more complicated expression, to add a private property that evaluates that expression and then use that property in the debugger attribute.  You can see this in a bunch of our types that have "DebuggerDisplay" properties.